### DISCUSSION TITLE
Introspecting `authenticated` tables to GraphQL schema?

### DISCUSSION THREAD

**Discussion by prescience-data** [supabase/supabase#6325] [Questions] [answered]
I'm trying to work out how to have `codegen` resolve my tables which are restricted to `authenticated` users when generating a schema.

So far I've tried:
1. using the `service_role` token as the `apiKey` header;
2. using a current `jwt` token from an `authenticated` session in the `authorization` header;
3. both of the above together;

Not sure what other angles to try, so interested if there's a way that I've missed! 👀

@olirice 

### Additional Context

Interestingly the `FooInsertResponse` items are generated (where `foo` table is protected), which seems like a possible security bug if users can reason about protected tables indirectly via these shapes.

**Comment by olirice** [comment]
> resolve my tables which are restricted to authenticated users when generating a schema.

For this you'll want to pass headers for your public `apiKey` & `Authorization: Bearer <jwt for authenticated user>`.

For example, on a sample project: here's the schema for an `anon` user (no auth header)
<img width="926" alt="anon" src="https://user-images.githubusercontent.com/12958657/161963869-159a5e46-10ef-4131-aa75-d328f3910d94.png">

and for an authenticated user
<img width="926" alt="authenticated" src="https://user-images.githubusercontent.com/12958657/161963898-e2aefe4f-2ea6-4729-8386-3f42873bde27.png">

> using the service_role token as the apiKey header

Thanks for letting me know. I was able to reproduce this error and am tracking it here https://github.com/supabase/pg_graphql/issues/148


> FooInsertResponse items are generated (where foo table is protected)

Thanks again, thats a bug. tracking here https://github.com/supabase/pg_graphql/issues/147
As a side note, that type only leaks the name of the table, not any field info. In case it worried you, even if a user were to attempt to use that type, it would be rejected by postgres' role based security since `pg_graphql` does not use escalated privileges (it uses the role from the JWT)

**Reply by olirice** [reply]
following up on 

> FooInsertResponse items are generated (where foo table is protected)

PR resolving the issue https://github.com/supabase/pg_graphql/pull/149

**Reply by prescience-data** [reply]
@olirice 

> _For this you'll want to pass headers for your public apiKey & Authorization: Bearer <jwt for authenticated user>._

I'm not able to reproduce this -

I am using the public api key provided by the dashboard, and a current authorization jwt provided by the `supabase-js` lib, however I am only seeing `anon` tables returned (which seems different to the behaviour in your screenshots).

The only way I can get the full table list is to grant `SELECT` to `anon`.

**Reply by olirice** [reply]
The issue you're seeing looks like its specific to codegen (possibly not passing a header in the expected way) vs an issue with `supabase` or `pg_graphql`. Lets continue figuring this out over at the issue you opened https://github.com/supabase/pg_graphql/issues/152 and loop in someone with experience using that tool

**Comment by prescience-data** [comment, marked as answer]
For others reference, the issue in this case was that I was inserting a colon in `Bearer: <token>` which had the effect of creating an invalid header.

@olirice 

For easier debugging, might be a good idea to throw an error if a user sends a malformed `Authorization` header rather than falling back to `anon` without feedback?
